libs/pageserver_api/src/shard.rs

```diff
@@ -526,6 +536,135 @@ impl ShardIdentity {
+
+        // Fast path: avoid hashing if we can immediately identify the owner of the whole range
+        if self.is_unsharded()
+            || self.count == ShardCount(1)
+            || (key_is_shard0(&start) && self.number == ShardNumber(0))
+        {
+            let start = (start.field5 as u64) << 32 | start.field6 as u64;
+            let end = (end.field5 as u64) << 32 | end.field6 as u64;
+
+            let diff = end - start;
+            if diff > u32::MAX as u64 {
+                return u32::MAX;
+            } else {
+                return diff as u32;
+            }
+        }
+
+        // Special cases for single keys like logical sizes
+        if end == start.add(1) && self.is_key_global(&start) {
+            return 1;
+        }
+
+        // Normal path: step through stripes and part-stripes in the range, evaluate whether each one belongs
+        // to Self, and add the stripe's block count to our total if so.
+        let mut result: u64 = 0;
+        let mut stripe_start = start;
+        while stripe_start < end {
+            let shard = key_to_shard_number(self.count, self.stripe_size, &stripe_start);
+
+            // Count up to the next stripe_size boundary
+            let stripe_index = stripe_start.field6 / self.stripe_size.0;
+            let stripe_remainder =
+                self.stripe_size.0 - (stripe_start.field6 - stripe_index * self.stripe_size.0);
+
+            let next_stripe_start = stripe_start.add(stripe_remainder);
+            let stripe_end = std::cmp::min(end, next_stripe_start);
+
+            // If this blocks in this stripe belong to us, add them to our count
+            if shard == self.number {
+                let start = (stripe_start.field5 as u64) << 32 | stripe_start.field6 as u64;
+                let end = (stripe_end.field5 as u64) << 32 | stripe_end.field6 as u64;
+                result += end - start;
+            }
+
+            stripe_start = next_stripe_start;
+        }
+        if result > u32::MAX as u64 {
+            u32::MAX
+        } else {
+            result as u32
+        }
+    }
+
+    /// Break up a key range into chunks, each of which has at least one local key in it.
+    pub fn key_range_chunk(
+        &self,
+        range: &Range<Key>,
+        target_nblocks: usize,
+    ) -> Vec<(u32, Range<Key>)> {
+        let mut range = range.clone();
+        let mut result = Vec::new();
+        loop {
+            // Split off the first target_nblocks if the remainder of the range would still contain
+            // some local blocks, otherwise yield the remainder of the range and we're done.
+            let range_size = self.key_range_local_size(&range);
+
+            if range_size > target_nblocks as u32 {
+                let remainder = Range {
+                    // FIXME: this add is not advancing far enough to capture target_nblocks *local*
+                    // blocks.  So we will end up chunking our range more finely than we needed to.
```

Comment:
This feels important since can be off multiples of stripe size, but I don't have a nice solution.

Reply:
I've fixed this by changing how the code works.  We now make one forward pass through the range, accumulating key range until we reach our target block count.